juju/juju
# Use sequences to track used charm URLs

Instead of relying on "dead" charm documents, which doesn't work well with migrations, local charm revisions are now tracked using a sequence. Dead charm documents are now removed.

An upgrade step is also included which adds any required sequences for pre-existing charms.

This PR also includes a fix for an unrelated gofmt issue in worker/machiner.

## QA steps

* Deploy using 2.1 branch
* Deploy a local charm
* Remove the local charm
* juju upgrade-juju to the code in this PR.
* Observe the the DB now contains the appropriate sequence & that the dead charm doc for the previous deploy is removed.
* Redeploy the local charm, observing the next charm revision is used.
* Deploy another application using the same local charm URL.
* Remove both applications, observing the charm is removed from the charms collection and the blobstore only once *both* applications are gone.
* Redeploy the local charm and observe that correct sequence is used.

## Bug reference

This is the final fix for https://bugs.launchpad.net/juju/+bug/1657614
